lib/plug/cowboy/drainer.ex

```diff
@@ -0,0 +1,119 @@
+defmodule Plug.Cowboy.Drainer do
+  @moduledoc """
+  Process to drain cowboy connections at shutdown.
+
+  When starting `Plug.Cowboy` in a supervision tree, it will create a listener that receives
+  requests and creates a connection process to handle that request. During shutdown, a
+  `Plug.Cowboy` process will immediately exit, closing the listener and any open connections
+  that are still being served. In most cases, it is desireable to allow connections to
+  complete before shutting down.
+
+  This module provides a process that during shutdown will close listeners and wait
+  for connections to complete. It should be placed after other supervised processes that
+  handle cowboy connections.
+
+  ## Options
+
+  The following options can be given to the child spec:
+
+    * `:refs` - A list of refs to drain. `:all` is also supported and will drain all cowboy
+      listeners, including those started by means other than `Plug.Cowboy`.
+
+    * `:id` - The ID for the process.
+      Defaults to `Plug.Cowboy.Drainer`.
+
+    * `:shutdown` - How long to wait for connections to drain.
+      Defaults to 5000ms.
+
+    * `:drain_check_interval` - How frequently to check if a listener's
+      connections have been drained.
+      Defaults to 1000ms.
+
+  ## Examples
+
+      # In your application
+      def start(_type, _args) do
+        children = [
+          {Plug.Cowboy, scheme: :http, plug: MyApp, options: [port: 4040]},
+          {Plug.Cowboy, scheme: :https, plug: MyApp, options: [port: 4041]},
+          {Plug.Cowboy.Drainer, refs: [MyApp.HTTP, MyApp.HTTPS]}
+        ]
+
+        opts = [strategy: :one_for_one, name: Hello.Supervisor]
+        Supervisor.start_link(children, opts)
+      end
+  """
+  use GenServer
+
+  @doc false
+  @spec child_spec(opts :: Keyword.t()) :: Supervisor.child_spec()
+  def child_spec(opts) when is_list(opts) do
+    {spec_opts, opts} = Keyword.split(opts, [:id, :shutdown])
+
+    Supervisor.child_spec(
+      %{
+        id: __MODULE__,
+        start: {__MODULE__, :start_link, [opts]},
+        type: :worker
+      },
+      spec_opts
+    )
+  end
+
+  @doc false
+  def start_link(opts) do
+    opts
+    |> Keyword.fetch!(:refs)
+    |> validate_refs!()
+
+    GenServer.start_link(__MODULE__, opts)
+  end
+
+  @doc false
+  @impl true
+  def init(opts) do
+    Process.flag(:trap_exit, true)
+    {:ok, opts}
+  end
+
+  @doc false
+  @impl true
+  def terminate(_reason, opts) do
+    opts
+    |> Keyword.fetch!(:refs)
+    |> drain(Keyword.get(opts, :drain_check_interval, 1_000))
+  end
+
+  defp drain(:all, drain_check_interval) do
+    :ranch.info()
+    |> Enum.map(&elem(&1, 0))
+    |> drain(drain_check_interval)
+  end
+
+  defp drain(refs, drain_check_interval) do
+    refs
+    |> Enum.filter(&suspend_listener/1)
```

Comment:
I changed this part to not use `Stream.each/2`, because I think it was tripping people up.

Changing `Enum.filter/2` to `Stream.filter/2` introduced a subtle bug illustrated by the following example.

``` elixir
# initial code looked like
[1, 2, 3, 4] 
|> Stream.each(&IO.inspect(&1, label: "first")) 
|> Enum.filter(fn e -> 
  IO.inspect(e, label: "filter")
  rem(e, 2) == 0 
end) 
|> Enum.each(&IO.inspect(&1, label: "final"))

# first: 1
# filter: 1
# first: 2
# filter: 2
# first: 3
# filter: 3
# first: 4
# filter: 4
# final: 2
# final: 4

[1, 2, 3, 4] 
|> Stream.each(&IO.inspect(&1, label: "first")) 
|> Stream.filter(fn e -> # with Stream.filter
  IO.inspect(e, label: "filter")
  rem(e, 2) == 0 
end) 
|> Enum.each(&IO.inspect(&1, label: "final"))

# first: 1
# filter: 1
# first: 2
# filter: 2
# final: 2
# first: 3
# filter: 3
# first: 4
# filter: 4
# final: 4
```

The important thing was that all the finals happen at the end. Otherwise you'd end up waiting for connections to drain for one listener, before suspending the next.

Also, the new code doesn't need to send a get_status to ranch.

Reply:
Sweet!